pilopt/src/lib.rs

```diff
@@ -614,11 +748,32 @@ fn remove_constant_intermediate_columns<T: FieldElement>(pil_file: &mut Analyzed
     substitute_polynomial_references(pil_file, intermediate_polys);
 }
 
-/// Substitutes all references to certain polynomials by the given field elements.
+/// Substitutes all references to certain polynomials by the given non-shifted expressions.
 fn substitute_polynomial_references<T: FieldElement>(
     pil_file: &mut Analyzed<T>,
-    substitutions: Vec<((String, PolyID), BigUint)>,
+    substitutions: Vec<((String, PolyID), AlgebraicExpression<T>)>,
 ) {
+    let poly_id_to_name = pil_file
```

Comment:
Can you re-use `build_poly_id_to_definition_name_lookup`? Also should we move that to `Analyzed` next to `name_to_poly_id`?

Reply:
Moved.